spring-web/src/main/java/org/springframework/http/server/reactive/ReactorNetty2ServerHttpRequest.java

```diff
@@ -75,48 +76,23 @@ public ReactorNetty2ServerHttpRequest(HttpServerRequest request, Netty5DataBuffe
 
 	private static URI initUri(HttpServerRequest request) throws URISyntaxException {
 		Assert.notNull(request, "HttpServerRequest must not be null");
-		return new URI(resolveBaseUrl(request) + resolveRequestUri(request));
+		return new URI(request.scheme() + "://" + resolveHost(request) + resolveRequestUri(request));
 	}
 
-	private static URI resolveBaseUrl(HttpServerRequest request) throws URISyntaxException {
-		String scheme = getScheme(request);
-
+	private static CharSequence resolveHost(HttpServerRequest request) {
 		InetSocketAddress hostAddress = request.hostAddress();
 		if (hostAddress != null) {
-			return new URI(scheme, null, hostAddress.getHostString(), hostAddress.getPort(), null, null, null);
+			return NetUtil.toSocketAddressString(hostAddress);
 		}
 
-		CharSequence charSequence = request.requestHeaders().get(HttpHeaderNames.HOST);
-		if (charSequence != null) {
-			String header = charSequence.toString();
-			final int portIndex;
-			if (header.startsWith("[")) {
-				portIndex = header.indexOf(':', header.indexOf(']'));
-			}
-			else {
-				portIndex = header.indexOf(':');
-			}
-			if (portIndex != -1) {
-				try {
-					return new URI(scheme, null, header.substring(0, portIndex),
-							Integer.parseInt(header, portIndex + 1, header.length(), 10), null, null, null);
-				}
-				catch (NumberFormatException ex) {
-					throw new URISyntaxException(header, "Unable to parse port", portIndex);
-				}
-			}
-			else {
-				return new URI(scheme, header, null, null);
-			}
+		CharSequence header = request.requestHeaders().get(HttpHeaderNames.HOST);
+		if (header != null) {
+			return header;
```

Comment:
The "Host" header can contain host and port, see for example https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host. Also, parsing needs to take into account that the host maybe an IPv6 address. This is what the removed code [89-110] does, and that needs to remain.